File: system_app_front_end/lib/areas/objects/data/table_payload.dart
import '../../ui/app_color_palettes.dart';

/// Canonical table object payload: [rows] grid + optional [chart] quality.
///
/// Chart tables are still `objects.type = table`; insert “graph” is sugar that
/// turns [chart.enabled] on with a default 2×N grid.
class TableObjectPayload {
  TableObjectPayload._();

  static const defaultChartType = 'bar';

  static Map<String, dynamic> empty({int columns = 2}) => {
        'rows': [
          [
            for (var c = 0; c < columns; c++) {'text': ''},
          ],
        ],
      };

  /// Default chart table (former graph insert).
  ///
  /// When [hebrewLabels] is true (UI RTL / Hebrew mode), series labels use
  /// א, ב, … instead of A, B, ….
  static Map<String, dynamic> emptyChart({
    int columns = 2,
    bool hebrewLabels = false,
  }) {
    final hexes = AppColorPalettes.defaultChart.hexes;
    return {
      'rows': [
        [
          for (var c = 0; c < columns; c++)
            {
              'text': hebrewLabels
                  ? String.fromCharCode(0x05D0 + c) // א, ב, …
                  : String.fromCharCode(0x41 + c), // A, B, …
            },
        ],
        [
          for (var c = 0; c < columns; c++) {'text': '${c + 1}'},
        ],
      ],
      'chart': {
        'enabled': true,
        'chartType': defaultChartType,
        'colors': [
          for (var c = 0; c < columns; c++) hexes[c % hexes.length],
        ],
      },
    };
  }

  static Map<String, dynamic> fromRowStrings(List<List<String>> rows) => {
        'rows': [
          for (final row in rows)
            [
              for (final cell in row) {'text': cell},
            ],
        ],
      };

  static bool chartEnabled(Map<String, dynamic>? payload) {
    if (payload == null) return false;
    final chart = payload['chart'];
    if (chart is Map && chart['enabled'] == true) return true;
    // Legacy graph shape.
    return payload.containsKey('labels') || payload.containsKey('values');
  }

  /// Accept rows, or legacy labels/values → rows + chart.
  static Map<String, dynamic> normalize(Map<String, dynamic>? payload) {
    final raw = payload == null
        ? <String, dynamic>{}
        : Map<String, dynamic>.from(payload);

    if (raw['rows'] is List) {
      final rows = _normalizeRows(raw['rows'] as List);
      final out = <String, dynamic>{'rows': rows};
      final chart = raw['chart'];
      if (chart is Map) {
        out['chart'] = _normalizeChart(
          Map<String, dynamic>.from(chart),
          columnCount: rows.isEmpty ? 2 : (rows.first as List).length,
        );
      }
      return out;
    }

    final labels = _stringList(raw['labels']);
    final values = _stringList(raw['values']);
    var colors = _stringList(raw['colors']);
    if (colors.isEmpty && raw['color'] != null) {
      colors = ['${raw['color']}'];
    }
    final chartType =
        '${raw['chartType'] ?? raw['chart_type'] ?? defaultChartType}'.trim();
    final n = [labels.length, values.length, colors.length, 2]
        .reduce((a, b) => a > b ? a : b);
    final lab = [...labels, ...List.filled(n, '')].take(n).toList();
    final val = [...values, ...List.filled(n, '')].take(n).toList();
    final cols = colors.isEmpty
        ? <String>[]
        : [...colors, ...List.filled(n, '')].take(n).toList();

    return {
      'rows': [
        [
          for (final t in lab) {'text': t},
        ],
        [
          for (final t in val) {'text': t},
        ],
      ],
      'chart': {
        'enabled': true,
        'chartType': chartType.isEmpty ? defaultChartType : chartType,
        'colors': cols,
      },
    };
  }

  static List<List<Map<String, dynamic>>> rowsOf(Map<String, dynamic> payload) {
    final n = normalize(payload);
    final raw = n['rows'] as List;
    return [
      for (final row in raw)
        if (row is List)
          [
            for (final cell in row)
              {
                'text': cell is Map ? '${cell['text'] ?? ''}' : '$cell',
              },
          ],
    ];
  }

  static Map<String, dynamic>? chartOf(Map<String, dynamic>? payload) {
    final n = normalize(payload);
    final chart = n['chart'];
    if (chart is! Map) return null;
    return Map<String, dynamic>.from(chart);
  }

  /// Marker / agent tag preference: chart-on → graph, else table.
  static String pointerObjectType(Map<String, dynamic>? payload) =>
      chartEnabled(payload) ? 'graph' : 'table';

  static List<String> _stringList(dynamic raw) {
    if (raw is! List) return [];
    return [for (final x in raw) '$x'];
  }

  static List<List<Map<String, dynamic>>> _normalizeRows(List rows) {
    final parsed = <List<Map<String, dynamic>>>[];
    for (final row in rows) {
      if (row is! List) continue;
      parsed.add([
        for (final cell in row)
          {
            'text': cell is Map ? '${cell['text'] ?? ''}' : '$cell',
          },
      ]);
    }
    if (parsed.isEmpty) {
      return [
        [
          {'text': ''},
          {'text': ''},
        ],
      ];
    }
    final maxCols = parsed.map((r) => r.length).reduce((a, b) => a > b ? a : b);
    final cols = maxCols < 1 ? 2 : maxCols;
    return [
      for (final row in parsed)
        [
          for (var c = 0; c < cols; c++)
            c < row.length ? row[c] : {'text': ''},
        ],
    ];
  }

  static Map<String, dynamic> _normalizeChart(
    Map<String, dynamic> chart, {
    required int columnCount,
  }) {
    final colorsRaw = chart['colors'];
    final colors = colorsRaw is List
        ? [for (final c in colorsRaw) '$c']
        : <String>[];
    final type =
        '${chart['chartType'] ?? chart['chart_type'] ?? defaultChartType}'
            .trim();
    return {
      'enabled': chart['enabled'] != false,
      'chartType': type.isEmpty ? defaultChartType : type,
      'colors': colors,
    };
  }
}

File: system_app_back_end/areas/objects/services/table_payload.py
"""Canonical table object payload: rows grid + optional chart quality."""

from __future__ import annotations

from typing import Any


def _cell(text: str = "") -> dict[str, Any]:
    return {"text": str(text), "spans": []}


def empty_table_payload(*, columns: int = 2) -> dict[str, Any]:
    return {
        "rows": [[_cell() for _ in range(columns)]],
    }


def empty_chart_table_payload(*, hebrew_labels: bool = False) -> dict[str, Any]:
    """Default chart table. FE insert passes labels; hebrew_labels → א/ב."""
    labels = ("א", "ב") if hebrew_labels else ("A", "B")
    return {
        "rows": [
            [_cell(labels[0]), _cell(labels[1])],
            [_cell("1"), _cell("2")],
        ],
        "chart": {
            "enabled": True,
            "chartType": "bar",
            "colors": ["#37899E", "#58C4D8"],
        },
    }


def chart_enabled(payload: dict[str, Any] | None) -> bool:
    if not payload:
        return False
    chart = payload.get("chart")
    if isinstance(chart, dict):
        return bool(chart.get("enabled"))
    # Legacy graph shape still has labels/values.
    return "labels" in payload or "values" in payload


def normalize_table_payload(payload: dict[str, Any] | None) -> dict[str, Any]:
    """Accept rows, or legacy labels/values graph shape → rows + chart."""
    raw = dict(payload or {})
    if "rows" in raw and isinstance(raw.get("rows"), list):
        rows = _normalize_rows(raw["rows"])
        out: dict[str, Any] = {"rows": rows}
        chart = raw.get("chart")
        if isinstance(chart, dict):
            out["chart"] = _normalize_chart(chart, column_count=len(rows[0]) if rows else 2)
        return out

    labels = [str(x) for x in (raw.get("labels") or [])]
    values = [str(x) for x in (raw.get("values") or [])]
    colors_raw = raw.get("colors")
    if not colors_raw and raw.get("color"):
        colors_raw = [raw.get("color")]
    colors = [str(x) for x in (colors_raw or [])] if colors_raw else []
    chart_type = str(raw.get("chartType") or raw.get("chart_type") or "bar").strip() or "bar"

    n = max(len(labels), len(values), len(colors), 2)
    labels = (labels + [""] * n)[:n]
    values = (values + [""] * n)[:n]
    colors = (colors + [""] * n)[:n] if colors else []

    return {
        "rows": [
            [_cell(t) for t in labels],
            [_cell(t) for t in values],
        ],
        "chart": {
            "enabled": True,
            "chartType": chart_type,
            "colors": colors,
        },
    }


def _normalize_rows(rows: list) -> list[list[dict[str, Any]]]:
    parsed: list[list[dict[str, Any]]] = []
    for row in rows:
        if not isinstance(row, list):
            continue
        parsed.append(
            [
                _cell(str(cell.get("text") if isinstance(cell, dict) else cell or ""))
                for cell in row
            ]
        )
    if not parsed:
        return [[_cell(), _cell()]]
    max_cols = max(len(r) for r in parsed)
    if max_cols < 1:
        max_cols = 2
    return [
        (row + [_cell()] * (max_cols - len(row)))[:max_cols]
        for row in parsed
    ]


def _normalize_chart(chart: dict[str, Any], *, column_count: int) -> dict[str, Any]:
    colors = chart.get("colors") or []
    if not isinstance(colors, list):
        colors = []
    colors = [str(c) for c in colors]
    return {
        "enabled": bool(chart.get("enabled", True)),
        "chartType": str(chart.get("chartType") or chart.get("chart_type") or "bar").strip()
        or "bar",
        "colors": colors,
    }


def rows_to_labels_values(payload: dict[str, Any]) -> tuple[list[str], list[str]]:
    """Agent GRAPH fence: first two rows as labels/values."""
    rows = payload.get("rows") or []
    if not rows:
        return [""], [""]
    labels = [
        str(c.get("text") if isinstance(c, dict) else c or "")
        for c in (rows[0] if isinstance(rows[0], list) else [])
    ]
    values = (
        [
            str(c.get("text") if isinstance(c, dict) else c or "")
            for c in (rows[1] if len(rows) > 1 and isinstance(rows[1], list) else [])
        ]
        if len(rows) > 1
        else [""] * len(labels)
    )
    n = max(len(labels), len(values), 1)
    labels = (labels + [""] * n)[:n]
    values = (values + [""] * n)[:n]
    return labels, values


def chart_meta(payload: dict[str, Any]) -> dict[str, Any]:
    chart = payload.get("chart") if isinstance(payload.get("chart"), dict) else {}
    colors = chart.get("colors") or []
    if not isinstance(colors, list):
        colors = []
    return {
        "chartType": str(chart.get("chartType") or "bar"),
        "colors": [str(c) for c in colors],
    }
